src/plugins/spec/spec-new.c

```diff
@@ -0,0 +1,754 @@
+			replaceCharacter (wildcardSpec, arraySpecKey, '_', '#');
+
+			char * formattedLookupKeyName = elektraFormat ("spec:/%s", arraySpecKey);
+			Key * foundKey = ksLookupByName (specKeys, formattedLookupKeyName, 0);
+			elektraFree (arraySpecKey);
+			elektraFree (formattedLookupKeyName);
+			if (foundKey != 0)
+			{
+				return foundKey;
+			}
+		}
+		else if (isArraySpecification (current))
+		{
+			const char * arraySpec = strchr (keyName (current), '/');
+
+			char * wildcardSpecKey = elektraMalloc (elektraStrLen (arraySpec));
+			replaceCharacter (arraySpec, wildcardSpecKey, '#', '_');
+
+			char * formattedLookupKeyName = elektraFormat ("spec:/%s", wildcardSpecKey);
+			Key * foundKey = ksLookupByName (specKeys, formattedLookupKeyName, 0);
+			elektraFree (wildcardSpecKey);
+			elektraFree (formattedLookupKeyName);
+			if (foundKey != 0)
+			{
+				return foundKey;
+			}
+		}
+	}
+	return 0;
+}
+
+static Key * getMatchingKeyFromKeySet (KeySet * ks, char * name)
+{
+	for (elektraCursor it = 0; it < ksGetSize (ks); it++)
+	{
+		Key * current = ksAtCursor (ks, it);
+		if (elektraStrCmp (strchr (keyName (current), '/'), name) == 0)
+		{
+			return current;
+		}
+	}
+	return 0;
+}
+
+/**
+ * Get the key which contains the array size as meta key.
+ *
+ * @param specKeys the specification keys
+ * @param specKey the specification key to use for lookup
+ * @return 0 - if the key was not found
+ * 	   key - the key which contains the array size
+ */
+static Key * getArraySizeOfArrayParent (KeySet * specKeys, Key * specKey)
+{
+	const char * specKeyName = keyName (specKey);
+	char * copiedKeyName = elektraMalloc (elektraStrLen (specKeyName) + 1);
+	strcpy (copiedKeyName, (char *) specKeyName);
+
+	char * arrayParent = strtok (copiedKeyName, "#");
+	arrayParent[strlen (arrayParent) - 1] = '\0';
+
+	Key * key = getMatchingKeyFromKeySet (specKeys, strchr (arrayParent, '/'));
+
+	elektraFree (copiedKeyName);
+
+	return key;
+}
+
+/**
+ * Get the number of array characters in an array specification key name.
+ *
+ * @param specKey the specification key to use
+ * @return number of array characters in the array specification key name
+ */
+static int getNumberOfArrayCharactersInSpecName (Key * specKey)
+{
+	char * withoutNamespace = strchr (keyName (specKey), '/');
+
+	int count = 0;
+
+	for (size_t i = 0; i < elektraStrLen (withoutNamespace); i++)
+	{
+		if (withoutNamespace[i] == '#')
+		{
+			count++;
+		}
+	}
+
+	return count;
+}
+
+/**
+ * The positions of where the `#` occurs in the specification key.
+ *
+ * @param keyNameWithoutNamespace the key name without the namespace
+ * @param arrayPositions the array to hold all the positions of the `#` in the specification key
+ * @param arraySize number of `#` in the whole specification key
+ */
+static void setArrayPositions (const char * keyNameWithoutNamespace, int * arrayPositions)
+{
+	int arrPos = 0;
+	for (size_t i = 0; i < elektraStrLen (keyNameWithoutNamespace); i++)
```

Comment:
```suggestion
	for (int i = 0; i < elektraStrLen (keyNameWithoutNamespace); i++)
```

Just use `int` then you don't need to cast below.

Reply:
Problem here is, that `elektraStrLen` returns `size_t`. I would need to cast it anyway, either once in the loop condition or each time on assignment. Probably just casting in the loop condition is "better" than doing it on each assignment, so I will do it there.

Reply:
No you don't need to cast. The compiler will automatically upcast the `int` to `size_t` for the comparison.

Reply:
Hmm, I get the following `error: comparison of integers of different signs: 'int' and 'size_t'`.